### DISCUSSION TITLE
Conversations with AI

### DISCUSSION THREAD

**Discussion by thorwhalen** [thorwhalen/umpyre#61] [General]
This discussion is to record some conversations I have with AI (for now, mainly chatGPT) about various python and software matters.

**Remember that AI is _only_ hallucination!** It's sometimes correct, but don't trust a thing it says. Look it up and try it out (if it seems safe). 

# Links to full chats:

* [Extending the LDA dimensional limit](https://chatgpt.com/share/67728eed-0894-8013-b55d-d7eb77868c59): Discusses ways to overcome the `n_components < n_unique_classes` limit of Linear Discriminant Analysis.

**Comment by thorwhalen** [comment]
## List package dependencies

**Does pip have an option or is there a python function that will give me the list of package dependencies of a given `pkg` package?**

Yes, there are several ways to get the list of package dependencies of a given package:

1. **Using pip**:
   The command `pip show` can be used to view the information about a particular package, which includes its dependencies.
   ```bash
   pip show <pkg-name>
   ```
   This will display the package's metadata, and you can look under the "Requires" field to see its dependencies.

2. **Using `pkg_resources`**:
   The `pkg_resources` module, which comes with `setuptools`, can be used to retrieve a package's dependencies programmatically:
   ```python
   import pkg_resources

   def get_dependencies(pkg_name):
       distribution = pkg_resources.get_distribution(pkg_name)
       return distribution.requires()

   print(get_dependencies("<pkg-name>"))
   ```

3. **Third-party Tools**: Tools like `pipdeptree` can be used to visualize the dependency tree of a package:
   First, you need to install `pipdeptree`:
   ```bash
   pip install pipdeptree
   ```
   Then, you can use it as follows to get dependencies of a specific package:
   ```bash
   pipdeptree --freeze --packages <pkg-name>
   ```

Choose the method that best fits your use case. If you just want to quickly check dependencies, using `pip` directly might be sufficient. If you need to do this programmatically in a script, then `pkg_resources` might be the way to go. If you need a more detailed and visual representation, `pipdeptree` could be helpful.

## ... do so without installing package?

**Can see the list of dependencies without installing the package, so I can check it out before pip installing?**

Yes, you can view the dependencies of a package without installing it using a few different methods:

1. **PyPI Website**:
   You can go to the [PyPI website](https://pypi.org/), search for the package, and under the "Meta" section on the right-hand side, you'll see a list of dependencies. Note, however, that this will show only direct dependencies and not the full recursive list of dependencies.

2. **Using `pip show` with a wheel file**:
   If you have a `.whl` (wheel) file for the package, you can use `pip show` on the wheel file to view its metadata, including dependencies:
   ```bash
   pip show /path/to/package_name-version.whl
   ```

3. **Using `pipdeptree` with a virtual environment**:
   This method involves creating a temporary virtual environment, installing the package there (which will isolate it from your global environment), and then using `pipdeptree` to view the dependencies.
   
   ```bash
   python -m venv mytempenv
   source mytempenv/bin/activate  # On macOS/Linux
   mytempenv\Scripts\activate  # On Windows
   
   pip install pipdeptree <pkg-name>
   pipdeptree --freeze --packages <pkg-name>
   
   deactivate
   rm -rf mytempenv  # or rmdir /s mytempenv on Windows
   ```
   
4. **Using `yolk3k`**:
   [`yolk3k`](https://pypi.org/project/yolk3k/) is a third-party tool that allows you to query PyPI and Python projects about their dependencies and other details without installing them.

   First, install `yolk3k`:
   ```bash
   pip install yolk3k
   ```

   Then, to list the dependencies:
   ```bash
   yolk -D <pkg-name>
   ```

The method you choose might depend on your exact requirements and how often you need to perform such checks. If it's a one-time thing, the PyPI website is probably the easiest. If you're doing this regularly and programmatically, one of the other methods might be more suitable.

**Reply by thorwhalen** [reply]
`yolk3k` hasn't been updated since 2016 at the time of writing this. 

And the `yolk -D <pkg-name>` didn't give dependencies.

**Reply by thorwhalen** [reply]
None of these worked in fact, so I wrote my own tool. See here: [unbox.dependencies_from_pypi](https://github.com/i2mint/unbox/blob/eb2f7f3f5a5745b86f95e2022feb235142c445cb/unbox/dependencies.py#L37). 

But then I noticed that though it all packages I've tried, it didn't work with any of the i2mint packages I've tried. 

So my conversation went further...

**Reply by thorwhalen** [reply]
### requires_dist empty?

**In one of my packages, it shows  'requires_dist': None,
yet I know I have dependencies and these are declared in setup.cfg. Why are they not showing up in the reqires_dist json of pypi and how can I make it so?**

If the `requires_dist` field is `None` for your package on PyPI, it usually means that the metadata for dependencies wasn't provided or wasn't successfully uploaded during the package distribution.

Here are a few common reasons and their solutions:

1. **Older Packaging Tool**:
   If you're using older packaging tools, they might not recognize or correctly use the `setup.cfg` file for dependencies. Ensure you're using a recent version of `setuptools`.

   Upgrade `setuptools` and `wheel`:
   ```bash
   pip install --upgrade setuptools wheel
   ```

2. **Metadata in `setup.cfg`**:
   Ensure that the dependencies are correctly specified in the `setup.cfg` under `[options]`. It should look something like this:
   ```cfg
   [options]
   install_requires =
       somepackage >= 1.0
       anotherpackage < 2.0
   ```

3. **Build and Upload with Modern Tools**:
   Consider using `build` to build the package and `twine` to upload it. This combination ensures that the package metadata is correctly built and uploaded to PyPI.

   Install the tools:
   ```bash
   pip install build twine
   ```

   Build your package:
   ```bash
   python -m build
   ```

   Upload using `twine`:
   ```bash
   twine upload dist/*
   ```

4. **Inspect the Built Package**:
   Before uploading, you can inspect the metadata of the built package. If you're building a wheel, the metadata is present in the `.dist-info` directory inside the wheel (which is a zip file). You can unzip the wheel and check the `METADATA` file.

5. **Other Metadata Files**: 
   If you have a `setup.py` in addition to `setup.cfg`, ensure that the `setup.py` does not inadvertently overwrite or omit the dependencies. The same caution applies if you have both `pyproject.toml` and `setup.cfg`.

6. **Check PyPI for Latest Version**:
   Ensure you're checking the correct version on PyPI. If you made changes in a new version but are looking at an older version on PyPI, you won't see the updated metadata.

If you've tried all the above and are still facing the issue, consider checking any open issues or quirks in the packaging tools you're using, or in PyPI itself. Sometimes, specific combinations of tools or unexpected scenarios can lead to odd behavior.

**Reply by thorwhalen** [reply]
### No wheel files

**When I check on the files section of my package on [pypi I see no wheel files](https://pypi.org/project/meshed/#files). I only see a 
tar.gz file.**

If you only see a `.tar.gz` file (source distribution) for your package on PyPI and no `.whl` files (wheel distribution), then it means you've only uploaded the source distribution and not the built wheel.

A wheel distribution (`.whl`) is a built, binary package format that allows for faster installations without needing to recompile the package (if it contains compiled code). For pure Python packages, wheel distributions ensure that the user doesn't require `setuptools` or other build tools to be installed when installing your package.

To correct this:

1. **Ensure you have the necessary tools**:
   Make sure you have `wheel` installed. If not, install it:
   ```bash
   pip install wheel
   ```

2. **Build the Wheel**:
   If you're using `setup.py`, you can build both source and wheel distributions with:
   ```bash
   python setup.py sdist bdist_wheel
   ```
   Alternatively, if you're using modern packaging with `pyproject.toml`, you can use:
   ```bash
   python -m build
   ```
   This will create both `.tar.gz` and `.whl` files in the `dist` directory.

3. **Upload to PyPI**:
   Use `twine` to upload both distributions:
   ```bash
   twine upload dist/*
   ```

After doing this, you should see both the `.tar.gz` (source distribution) and `.whl` (wheel distribution) files for your package on PyPI.

Remember, if your package contains compiled code (e.g., C extensions), you may need to build wheels for specific platforms or Python versions. Tools like `cibuildwheel` can assist in building wheels for multiple platforms. For pure Python code, a single universal wheel should suffice.

**Comment by thorwhalen** [comment]
## Feedback loop

_What is the law of ux that states that a user needs to have feedback on whether their action request was registered?_

Feedback Loop: The Feedback Loop is a UX law that emphasizes the importance of providing immediate feedback to users. According to this law, users should receive feedback that lets them know what actions they have taken, whether those actions were successful, and what steps they need to take next.

Some links:
[UX: Creating natural experiences with feedback loops](https://medium.com/codeandco/ux-making-experiences-feel-natural-with-feedback-loops-ce4eb333f99c)
[24 Essential UX Laws for a Great Product Design](https://www.onething.design/blogs/laws-of-ux/#:~:text=22.%20Feedback%20Loop,and%20error%20messages.)

### Gemini

The Feedback Loop is a core principle in UX design that emphasizes the importance of keeping users informed through clear and timely responses. Here are some key aspects of this law:

Immediate Feedback: Every user action should be acknowledged with a prompt response. This can be visual cues like a button changing state, sounds, or haptic feedback on mobile devices. Essentially, the user needs to feel their action is recognized by the system.

Clear Communication: Feedback should effectively communicate the outcome of the user's action. Did the button press register? Was the form submitted successfully? Error messages should be clear and actionable, guiding the user towards a solution.

Progress Updates: For longer processes, keep users informed about the progress. Progress bars, percentage completion indicators, or estimated wait times can all help manage user expectations and avoid frustration.

Natural Timing: Feedback should be delivered at a natural pace. Avoid overly slow or instantaneous responses that can be jarring to the user.

Placement: Feedback needs to be positioned where users expect to see it. For instance, a confirmation message following a button press should appear near the button itself, not hidden elsewhere on the screen.

By following these principles, you can create a UX that feels responsive and intuitive, fostering trust and satisfaction in your users.

**Comment by thorwhalen** [comment]
## Leaky Lambdas in loops

In this code:

```python
def extension_based_decoding_store_wrappers(extension_to_decoder):
    for ext, decoder in extension_to_decoder.items():
        yield Pipe(
            filt_iter(filt=lambda x: x.endswith(ext)), 
            wrap_kvs(value_decoder=decoder)
        )
```

I end up having, in each Pipe yielded, the same exact filt_iter object, using the same ext -- the last one that was seen in the for loop. 

I know it has to do with using the lambda function, but don't know exactly why.

Explain.

### chatGPT (Python GPT)

The issue you're encountering is related to the way Python's closures capture variables in lambda functions within loops. This is a common pitfall when using lambda functions inside loops, especially when the lambda is intended to capture the value of a loop variable at each iteration.

In the code snippet you provided, the lambda function `lambda x: x.endswith(ext)` captures the variable `ext`. However, it doesn't capture the value of `ext` at each iteration of the loop. Instead, it captures the variable itself, not its current value during each iteration. Therefore, by the time the lambda function is executed, it uses the value of `ext` present at the last iteration of the loop for all created `Pipe` objects, not the value of `ext` at the time the lambda was defined.

This behavior is due to how closures work in Python: a closure captures variables, not their values. The captured variables are resolved when the function is called, not when it's defined. Thus, all lambda functions defined in the loop capture and share the same `ext` variable, which after the loop ends, retains its final value from the loop.

### Solution

A common solution to this issue is to use a default argument in the lambda function that captures the current value of the loop variable at each iteration. By doing this, you ensure that each lambda function has its own copy of the value, as default argument values are evaluated at function definition time, not at execution time.

Here’s how you can modify the lambda function to correctly capture the value of `ext` at each iteration:

```python
def extension_based_decoding_store_wrappers(extension_to_decoder):
    for ext, decoder in extension_to_decoder.items():
        yield Pipe(
            filt_iter(filt=lambda x, ext=ext: x.endswith(ext)),  # Use default argument to capture the loop variable
            wrap_kvs(value_decoder=decoder)
        )
```

By adding `ext=ext` as a default argument to the lambda, each lambda function now has its own `ext` value that corresponds to the loop iteration during which the lambda was created. This solves the problem and ensures that each `Pipe` object uses the correct `ext` value.

**Comment by thorwhalen** [comment]
## Convention over configuration and configuration as code

_In declarative approaches of software, code is moved to configuration and configuration to convention. One kind of convention has to do with rules to eliminate frequent configuration patterns through defaults or rules that enable the user to write less explicit configuration._

_Tell me about software design principles and design patterns that are related to this._

The principle of moving code to configuration and further reducing explicit configuration through conventions is aligned with several key software design principles and design patterns. These concepts aim to improve code readability, maintainability, scalability, and efficiency by promoting more abstract, high-level, and declarative coding styles. Here are some of the most relevant principles and patterns related to this approach:

### Software Design Principles

1. **Convention Over Configuration (CoC):**
   - This principle minimizes the number of decisions that developers need to make, gaining simplicity but not necessarily losing flexibility. The idea is that a system or framework can assume reasonable defaults about what a developer intends based on conventions, thereby reducing the amount of configuration required.

2. **Don't Repeat Yourself (DRY):**
   - The DRY principle emphasizes the need to reduce repetition of software patterns. Moving repetitive code to a shared configuration or utilizing conventions effectively reduces the need to duplicate similar configurations across different parts of the application.

3. **Configuration as Code (CaC):**
   - This approach treats configuration files, which dictate how an application operates, as if they were code. This means they are written in a high-level language, version-controlled, and subject to the same review processes as application code. This allows for more dynamic and flexible application behavior based on declarative configurations.

4. **Separation of Concerns (SoC):**
   - SoC is about separating a computer program into distinct sections, such that each section addresses a separate concern. By separating concerns into configuration and convention, the core logic of the application is decoupled from configuration logic, leading to cleaner and more maintainable code.

### Design Patterns

1. **Strategy Pattern:**
   - This pattern enables the selection of algorithms at runtime. Similarly, conventions can be seen as strategies for configuration, allowing different behaviors or configurations to be selected based on certain conventions or rules.

2. **Factory Pattern:**
   - Factories abstract the process of object creation and allow the creation process to be configured or extended. In a way, moving code to configuration can leverage factory-like behavior, where configurations dictate the creation and behavior of objects in a flexible manner.

3. **Template Method Pattern:**
   - This pattern defines the program skeleton of an algorithm in a method, deferring some steps to subclasses. It's akin to defining a convention (the skeleton) while allowing specifics to be configured or automatically assumed (the subclass implementations).

4. **Interpreter Pattern:**
   - This pattern is about defining a representation for its grammar along with an interpreter that uses the representation to interpret sentences in the language. In a broad sense, conventions can be seen as a grammar, and the system or framework acts as the interpreter, understanding and executing based on those conventions.

These principles and patterns demonstrate a shift towards more declarative, abstract, and high-level approaches in software development, focusing on what needs to be achieved rather than how it should be done explicitly. This not only reduces the boilerplate code but also makes the codebase more adaptable and easier to maintain.

### A discussed example

_Is there a principle or design pattern that has to do with factoring out most common configuration values as conventional defaults?_

_For example, if I note that a "method" field's value is most often "GET", I can most this information to a "defaults" part of the convention, thereby being able to omit the method field all together, except when the "method" field's value is not "GET"._

_Tell me about this, the lingo around it, and the pros and cons._

The principle closely related to factoring out common configuration values as conventional defaults is **Convention Over Configuration (CoC)**, also sometimes referred to as **Coding by Convention**. This principle is particularly prevalent in modern software frameworks and environments, aiming to simplify development by minimizing the number of decisions developers need to make and reducing the amount of boilerplate code they need to write.

#### Convention Over Configuration Explained

In the context of Convention Over Configuration, defaults are established based on the most common scenarios or use cases. These defaults are automatically assumed by the system unless explicitly overridden by the developer. This approach is exactly as you described: if a "method" field in an API call is most often "GET", the system assumes "GET" as the default method unless a different method is specified. This reduces the need for repetitive and redundant configuration, making code cleaner and development faster.

#### Terminology

- **Defaults**: Predefined values or settings assumed by the system in the absence of explicit configurations.
- **Conventions**: Established patterns and practices that are automatically assumed or applied by the system.
- **Configuration**: Specific parameters or settings provided by the developer to modify the default behavior of the system.

#### Pros of Convention Over Configuration

1. **Simplified Development**: Reduces the amount of configuration code developers need to write, making the development process faster and less error-prone.
2. **Enhanced Readability**: By removing boilerplate configuration, the remaining code is more focused on the unique aspects of the application, improving readability.
3. **Easier Maintenance**: With fewer configurations to manage, maintaining and updating the application becomes simpler.
4. **Rapid Prototyping**: Allows for quicker iterations and prototyping, as the default configurations handle most of the setup automatically.

#### Cons of Convention Over Configuration

1. **Learning Curve**: Developers need to learn the conventions used by the framework or system, which can be a barrier for new users.
2. **Less Flexibility**: While defaults can simplify development, they might also limit flexibility by making it harder to deviate from the conventional paths.
3. **Implicit Behavior**: The automatic application of conventions can lead to unexpected behavior if developers are not fully aware of the underlying conventions.
4. **Overhead for Simple Tasks**: In some cases, setting up a system that heavily relies on conventions can introduce overhead for tasks that deviate from the most common scenarios.

In summary, Convention Over Configuration is a powerful principle for streamlining development by leveraging sensible defaults and established conventions. While it offers significant benefits in terms of simplification and efficiency, it's important for developers to be well-versed in the conventions used by their chosen frameworks or systems to fully leverage this approach and avoid potential pitfalls.

**Comment by thorwhalen** [comment]
## Markdown hyperlinked TOC don't work on pypi

By this I mean, table of contents that use this format:

```markdown
- [Usage](#usage)
- [Background](#background)
```

I thought perhaps it was only the way I did things myself that made it work on github, but not pypi.
Then I checked out others. See for example `pandas`: [github](https://github.com/pandas-dev/pandas/blob/main/README.md) TOC links do jump to the relevant section, but the same readme, when [viewed in pypi](https://pypi.org/project/pandas/), don't work. 
Click on `background` for example will change the url to https://pypi.org/project/pandas/#background. 

Now, given that even `pandas` TOC "doesn't work", I'm guessing there's no easy solution, and it must be something that pypi should do on their side. But I thought I'd check here in case there was something we're all missing. 

*Answer*

The issue you've encountered with Markdown TOC (Table of Contents) links not working on PyPI is a known limitation of the PyPI platform. This behavior occurs because PyPI does not fully support the fragment identifiers (the `#` links) that are commonly used in Markdown for navigating to sections within the same document.

### Background on the Issue

Markdown allows you to create internal links using fragment identifiers, which work well on platforms like GitHub that render Markdown and generate HTML anchors automatically for each heading. For instance, on GitHub, the heading `## Usage` will automatically be assigned an `id` attribute like `#usage`, making it possible to link to it using `[Usage](#usage)`.

However, PyPI's rendering engine for Markdown, which is part of the package uploading process, does not create the necessary HTML `id` attributes for headings. As a result, while the links in the TOC on PyPI will update the URL in the browser to include the fragment (e.g., `#background`), the browser cannot jump to the section because the corresponding `id` does not exist in the rendered HTML.

### Is There a Solution?

As of now, there is no direct fix or workaround on the user side to make these internal links work on PyPI, since this is a limitation of how PyPI renders Markdown. The underlying issue would need to be addressed by PyPI itself, either by enhancing its Markdown renderer to automatically generate `id` attributes for headings or by supporting a different syntax that allows users to specify `id` attributes directly.

### Current Best Practices

Given this limitation, here are some recommendations:

1. **Minimize reliance on internal links in your PyPI README**: If possible, avoid relying on internal links for navigation within your README when it's rendered on PyPI. Instead, consider summarizing sections or linking to detailed documentation hosted elsewhere (e.g., on GitHub or a dedicated documentation site).

2. **External Documentation**: For more complex projects where internal linking within the README is essential, consider directing users to external documentation platforms like ReadTheDocs or a GitHub Wiki, where these links will work correctly.

3. **Raise Awareness**: If this is a significant concern for your project, consider raising the issue in the Python Packaging Authority's (PyPA) GitHub repository to see if there is interest or ongoing work to improve PyPI's Markdown rendering capabilities.

### Conclusion

The issue with TOC links not working on PyPI is due to limitations in PyPI's handling of Markdown. While GitHub correctly processes and renders these links, PyPI does not currently support the necessary HTML `id` attributes for internal links to function properly. As a result, there's no easy solution available on the user side, and addressing this limitation would require changes to PyPI's rendering engine.

### Links

Here are some links to relevant discussions and issues that address the problem of Markdown TOC links not working on PyPI:

1. **PyPA Warehouse Issue on GitHub**: This is where issues related to PyPI's backend, including Markdown rendering, are tracked.
   - [Markdown rendering issue on PyPI (GitHub)](https://github.com/pypa/warehouse/issues/746)

2. **Python Packaging Discourse**: This is a forum where the Python packaging community discusses various topics, including issues related to PyPI. You might find ongoing discussions or be able to start one here.
   - [Python Packaging Discourse](https://discuss.python.org/c/packaging)

3. **Markdown Support Discussion on PyPA Warehouse**: Another discussion that touches on Markdown rendering issues in general on PyPI.
   - [General Markdown rendering discussion on PyPI (GitHub)](https://github.com/pypa/warehouse/issues/5899)

These links should help you stay updated on any developments or community discussions regarding the issue. If you don't find a specific ongoing discussion about your exact issue, these platforms are also good places to raise awareness or report new observations.
